lnst/Recipes/ENRT/ConfigMixins/BaseRESTConfigMixin.py

```diff
@@ -28,11 +30,27 @@ def __build_request(self, endpoint: str, **kwargs):
 
         return kwargs
 
-    def api_request(self, method: str, endpoint: str, response_code: int = 200, **kwargs) -> bytes:
+    def _calculate_retry_delay(self, retry: int):
+        base_delay = 1
+        return base_delay * 2 ** retry
+
+    def api_request(self, method: str, endpoint: str, response_code: int = 200, tries: int = 7, **kwargs) -> bytes:
         request = self.__build_request(endpoint, **kwargs)
         req_func = self.__get_request_function(method)
 
-        response = req_func(**request)
+        for i in range(1, tries+1):  # i starts from 1
+            try:
+                response = req_func(**request)
+            except RequestException as e:
+                logging.error(f"API request ({i}/{tries}) failed: {e}")
+
+                time.sleep(self._calculate_retry_delay(i))
+                continue
+            
+            break  # request was successful
+        else:
+            raise LnstError(f"API request failed after {tries} tries")
+
         if response.status_code != response_code:
             raise LnstError(f"Request failed with status code {response.status_code}")
 
```

Comment:
To simplify code even further we could combine the try-except with status code check that you do after the block.
Also move base_delay somewhere, that isn't going to reinitialize it each call.

```python
try:
    response = req_func(**request)
    if response.status_code == response_code:
        logging.debug("API response: %s", response.content)
        return response.content
    else:
        logging.error(f"Request failed with status code {response.status_code} (attempt {i}/{tries})")
except RequestException as e:
    logging.error(f"API request (attempt {i}/{tries}) failed: {e}")

    delay = self.BASE_DELAY * 2 ** (i - 1)
    logging.info(f"Retrying in {delay} seconds...")
    time.sleep(delay)
```

Reply:
agree with the delay calculation proposed.

I probably wouldn't include the status code check inside the try-except block as it increases the indentation level and doesn't actually change anything runtime wise.

Unless we also want to do retries also in case the response code isn't the expected value? I'm not sure we want that universally (i can imagine situations when a bad status code means we shouldn't retry...) but it could make sense in some situations?

Reply:
The idea behind retrying the calls is, to handle more or less networking problems. Non-expected http return codes implies problems with request itself, so i don't think it should be retried

Reply:
yeah, that's probably true, agreed. let's just do the delay calculation proposal and we can merge this in.